mars/learn/linear_model/_base.py

```diff
@@ -0,0 +1,375 @@
+        positive=False,
+    ):
+        self.fit_intercept = fit_intercept
+        self.normalize = normalize
+        self.copy_X = copy_X
+        self.positive = positive
+
+    def fit(self, X, y, sample_weight=None):
+        """
+        Fit linear model.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+
+        y : array-like of shape (n_samples,) or (n_samples, n_targets)
+            Target values. Will be cast to X's dtype if necessary.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            Individual weights for each sample.
+
+        Returns
+        -------
+        self : object
+            Fitted Estimator.
+        """
+        accept_sparse = False if self.positive else ['csr', 'csc', 'coo']
+
+        X, y = self._validate_data(X, y, accept_sparse=accept_sparse,
+                                   y_numeric=True, multi_output=True)
+
+        if sample_weight is not None:
+            sample_weight = _check_sample_weight(sample_weight, X,
+                                                 dtype=X.dtype)
+
+        X, y, X_offset, y_offset, X_scale = self._preprocess_data(
+            X, y, fit_intercept=self.fit_intercept, normalize=self.normalize,
+            copy=self.copy_X, sample_weight=sample_weight,
+            return_mean=True)
+
+        if sample_weight is not None:
+            # Sample weight can be implemented via a simple rescaling.
+            X, y = _rescale_data(X, y, sample_weight)
+
+        if self.positive:
+            # # When optimize.nnls is implemented, one can directly
+            # # utilize the following code.
+            # if y.ndim < 2:
+            #     self.coef_, self._residues = optimize.nnls(X, y)
+            # else:
+            #     # scipy.optimize.nnls cannot handle y with shape (M, K)
+            #     outs = Parallel(n_jobs=n_jobs_)(
+            #         delayed(optimize.nnls)(
+            #             X, y[:, j]) for j in range(y.shape[1]))
+            #     self.coef_, self._residues = map(mt.vstack, zip(*outs))
+            #     self.coef_.execute()
+            raise NotImplementedError(
+                "Does not support positive coefficients!")
+        elif sp.issparse(X):
+            X_offset_scale = X_offset / X_scale
+
+            def matvec(b):
+                return X.dot(b) - b.dot(X_offset_scale)
+
+            def rmatvec(b):
+                return X.T.dot(b) - X_offset_scale * mt.sum(b)
+
+            X_centered = sp.linalg.LinearOperator(shape=X.shape,
+                                                  matvec=matvec,
+                                                  rmatvec=rmatvec)
+
+            if y.ndim < 2:
+                out = lsqr(X_centered, y)
+                self.coef_ = out[0]
+                self._residues = out[3]
+            else:
+                outs = []
+                for j in range(y.shape[1]):
```

Comment:
Mars now doesn't support  the `lsqr` function, if there's no workaround for it, we can raise a `NotImplementedError` here.

Reply:
The test case is a bit tricky, I might need a few more days.